// bhs/server/BHSAPI.js
var inherits = require('util').inherits;
var EventEmitter = require('events').EventEmitter;
var hap = require('./BHSAPHK');
var hapLegacyTypes = require('./BHSAT');
var log = require("./BHSLR")._system;
var User = require("./BHSU").User;
var PlatformAccessory = require("./BHSPA").PlatformAccessory;
var serverVersion = require("./BHSV");

function API() {
  this._accessories = {};
  this._platforms = {};
  this._configurableAccessories = {};
  this._dynamicPlatforms = {};  
  this.version = 2.2;
  this.serverVersion = serverVersion;
  this.user = User;
  this.hap = hap;
  this.hapLegacyTypes = hapLegacyTypes;
  this.platformAccessory = PlatformAccessory;
}

inherits(API, EventEmitter);

API.prototype.accessory = function(name) {
  
  if (name.indexOf('.') == -1) {
    var found = [];
    for (var fullName in this._accessories) {
      if (fullName.split(".")[1] == name)
        found.push(fullName);
    }
    
    if (found.length == 1) {
      return this._accessories[found[0]];
    }
    else if (found.length > 1) {
      throw new Error("The requested accessory '" + name + "' has been registered multiple times. Please be more specific by writing one of: " + found.join(", "));
    }
    else {
      throw new Error("The requested accessory '" + name + "' was not registered by any plugin.");
    }
  }
  else {

    if (!this._accessories[name])
      throw new Error("The requested accessory '" + name + "' was not registered by any plugin.");
    
    return this._accessories[name];
  }
}

API.prototype.registerAccessory = function(pluginName, accessoryName, constructor, configurationRequestHandler) {
  var fullName = pluginName + "." + accessoryName;
  
  if (this._accessories[fullName])
    throw new Error("Attempting to register an accessory '" + fullName + "' which has already been registered!");

  log.info("Registering accessory '%s'", fullName);
  this._accessories[fullName] = constructor;

  if (configurationRequestHandler) {
    this._configurableAccessories[fullName] = configurationRequestHandler;
  }
}

API.prototype.publishCameraAccessories = function(pluginName, accessories) {
  for (var index in accessories) {
    var accessory = accessories[index];
    if (!(accessory instanceof PlatformAccessory)) {
      throw new Error(pluginName + " attempt to register an accessory that isn\'t PlatformAccessory!");
    }
    accessory._associatedPlugin = pluginName;
  }

  this.emit('publishCameraAccessories', accessories);
}

API.prototype.platform = function(name) {
  
 if (name.indexOf('.') == -1) {
    var found = [];
    for (var fullName in this._platforms) {
      if (fullName.split(".")[1] == name)
        found.push(fullName);
    }
    
    if (found.length == 1) {
      return this._platforms[found[0]];
    }
    else if (found.length > 1) {
      throw new Error("The requested platform '" + name + "' has been registered multiple times. Please be more specific by writing one of: " + found.join(", "));
    }
    else {
      throw new Error("The requested platform '" + name + "' was not registered by any plugin.");
    }
  }
  else {

    if (!this._platforms[name])
      throw new Error("The requested platform '" + name + "' was not registered by any plugin.");
    
    return this._platforms[name];
  }
}

API.prototype.registerPlatform = function(pluginName, platformName, constructor, dynamic) {
  var fullName = pluginName + "." + platformName;
  
  if (this._platforms[fullName])
    throw new Error("Attempting to register a platform '" + fullName + "' which has already been registered!");

  log.info("Registering platform '%s'", fullName);

  this._platforms[fullName] = constructor;

  if (dynamic) {
    this._dynamicPlatforms[fullName] = constructor;
  }
}

API.prototype.registerPlatformAccessories = function(pluginName, platformName, accessories) {
  for (var index in accessories) {
    var accessory = accessories[index];
    if (!(accessory instanceof PlatformAccessory)) {
      throw new Error(pluginName + " - " + platformName + " attempt to register an accessory that isn\'t PlatformAccessory!");
    }
    accessory._associatedPlugin = pluginName;
    accessory._associatedPlatform = platformName;
  }

  this.emit('registerPlatformAccessories', accessories);
}

API.prototype.updatePlatformAccessories = function(accessories) {
  this.emit('updatePlatformAccessories', accessories);
}

API.prototype.unregisterPlatformAccessories = function(pluginName, platformName, accessories) {
  for (var index in accessories) {
    var accessory = accessories[index];
    if (!(accessory instanceof PlatformAccessory)) {
      throw new Error(pluginName + " - " + platformName + " attempt to unregister an accessory that isn\'t PlatformAccessory!");
    }
  }
  this.emit('unregisterPlatformAccessories', accessories);
}

module.exports = {
  API: API
};
