// bhs/server/BHSAR.js
'use strict';
var mdns = require('mdns');
var crypto = require('crypto');

function Advertiser(accessoryInfo) {
  this.accessoryInfo = accessoryInfo;
  this._advertisement = null;

  this._setupHash = this._computeSetupHash();
}

Advertiser.prototype.startAdvertising = function(port) {

  if (this._advertisement) {
    this.stopAdvertising();
  }

  var txtRecord = {
    md: this.accessoryInfo.displayName,
    pv: "1.0",
    id: this.accessoryInfo.username,
    "c#": this.accessoryInfo.configVersion + "",
    "s#": "1",
    "ff": "0",
    "ci": this.accessoryInfo.category,
    "sf": this.accessoryInfo.paired() ? "0" : "1",
    "sh": this._setupHash
  };

  this._advertisement = mdns.createAdvertisement(mdns.tcp('hap'), port, {
    name: this.accessoryInfo.displayName,
    txtRecord: txtRecord
  });
  this._advertisement.start();
}

Advertiser.prototype.isAdvertising = function() {
  return (this._advertisement != null);
}

Advertiser.prototype.updateAdvertisement = function() {
  if (this._advertisement) {

    var txtRecord = {
      md: this.accessoryInfo.displayName,
      pv: "1.0",
      id: this.accessoryInfo.username,
      "c#": this.accessoryInfo.configVersion + "",
      "s#": "1",
      "ff": "0",
      "ci": this.accessoryInfo.category,
      "sf": this.accessoryInfo.paired() ? "0" : "1",
      "sh": this._setupHash
    };

    this._advertisement.updateTXTRecord(txtRecord);
  }
}

Advertiser.prototype.stopAdvertising = function() {
  if (this._advertisement) {
    this._advertisement.stop();
    this._advertisement = null;
  }
}

Advertiser.prototype._computeSetupHash = function() {
  var setupHashMaterial = this.accessoryInfo.setupID + this.accessoryInfo.username;
  var hash = crypto.createHash('sha512');
  hash.update(setupHashMaterial);
  var setupHash = hash.digest().slice(0, 4).toString('base64');

  return setupHash;
}

module.exports = {
  Advertiser: Advertiser
};
