// bhs/server/BHSGS.js
'use strict';
var path = require('path');
var fs = require('fs');
var uuid = require('./BHSAPHK').uuid;
var accessoryStorage = require('node-persist').create();
var Bridge = require('./BHSAPHK').Bridge;
var Accessory = require('./BHSAPHK').Accessory;
var Service = require('./BHSAPHK').Service;
var Characteristic = require('./BHSAPHK').Characteristic;
var AccessoryLoader = require('./BHSAPHK').AccessoryLoader;
var once = require('./util/once').once;
var Plugin = require('./BHSP').Plugin;
var User = require('./BHSU').User;
var API = require('./BHSAPI').API;
var PlatformAccessory = require("./BHSPA").PlatformAccessory;
var BridgeSetupManager = require("./BHSBSM").BridgeSetupManager;
var log = require('./BHSLR')._system;
var Logger = require('./BHSLR').Logger;
var mac = require("./util/mac");
var chalk = require('chalk');
var qrcode = require('qrcode-terminal');

function toTitleCase(str) {
  return str.replace(/\w\S*/g, txt => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase());
}

function Server(insecureAccess, opts) {
  opts = opts || {};

  accessoryStorage.initSync({ dir: User.cachedAccessoryPath() });

  this._api = new API();

  this._api.on('registerPlatformAccessories', function(accessories) {
    this._handleRegisterPlatformAccessories(accessories);
  }.bind(this));

  this._api.on('updatePlatformAccessories', function(accessories) {
    this._handleUpdatePlatformAccessories(accessories);
  }.bind(this));

  this._api.on('unregisterPlatformAccessories', function(accessories) {
    this._handleUnregisterPlatformAccessories(accessories);
  }.bind(this));

  this._api.on('publishCameraAccessories', function(accessories) {
    this._handlePublishCameraAccessories(accessories);
  }.bind(this));

  this._plugins = this._loadPlugins();
  this._config = opts.config || this._loadConfig();
  this._cachedPlatformAccessories = this._loadCachedPlatformAccessories();
  this._bridge = this._createBridge();

  this._activeDynamicPlugins = {};
  this._configurablePlatformPlugins = {};
  this._publishedCameras = {};
  this._setupManager = new BridgeSetupManager();
  this._setupManager.on('newConfig', this._handleNewConfig.bind(this));

  this._setupManager.on('requestCurrentConfig', function(callback) {
    callback(this._config);
  }.bind(this));

  this._allowInsecureAccess = insecureAccess || false;
}

Server.prototype.run = function() {

  this._asyncCalls = 0;
  this._asyncWait = true;
  if (this._config.platforms) this._loadPlatforms();
  if (this._config.accessories) this._loadAccessories();
  this._loadDynamicPlatforms();
  this._configCachedPlatformAccessories();
  this._setupManager.configurablePlatformPlugins = this._configurablePlatformPlugins;
  this._bridge.addService(this._setupManager.service);
  this._asyncWait = false;

  if (this._asyncCalls == 0)
    this._publish();

  this._api.emit('didFinishLaunching');
}

Server.prototype._publish = function() {
  var bridgeConfig = this._config.server || {};
  var info = this._bridge.getService(Service.AccessoryInformation);
  info.setCharacteristic(Characteristic.Manufacturer, `${toTitleCase(require('../package.json').author.name)}`);
  info.setCharacteristic(Characteristic.Model, `${toTitleCase(require('../package.json').name)}`);
  info.setCharacteristic(Characteristic.SerialNumber, bridgeConfig.username);
  info.setCharacteristic(Characteristic.FirmwareRevision, require('../package.json').version);

  this._bridge.on('listening', function(port) {
    log.info("BioXHOME running in the port %s.", port);
  });

  var publishInfo = {
    username: bridgeConfig.username || "99:89:83:63:64:12",
    port: bridgeConfig.port || 0,
    pincode: bridgeConfig.pin || "836-36-412",
    category: Accessory.Categories.BRIDGE
  };

  if (bridgeConfig.setupID && bridgeConfig.setupID.length === 4) {
    publishInfo['setupID'] = bridgeConfig.setupID;
  }

  this._bridge.publish(publishInfo, this._allowInsecureAccess);
  this._printSetupInfo();
  this._printPin(publishInfo.pincode);
}

Server.prototype._loadPlugins = function(accessories, platforms) {

  var plugins = {};
  var foundOnePlugin = false;

  Plugin.installed().forEach(function(plugin) {

    try {
      plugin.load();
    }
    catch (err) {
      log.error("Error loading plugin " + plugin.name() + ":")
      log.error(err.stack);
      plugin.loadError = err;
    }

    if (!plugin.loadError) {

      plugins[plugin.name()] = plugin;
      log.info("Plugin loaded: " + plugin.name());
      plugin.initializer(this._api);
      log.info("---");
      foundOnePlugin = true;
    }

  }.bind(this));

  if (!foundOnePlugin) {
    log.warn("No plugins found. See the README for information on installing plugins.")
  }

  return plugins;
}

Server.prototype._loadConfig = function() {

  var configPath = User.configPath();

  if (!fs.existsSync(configPath)) {
    log.warn("server.json (%s) not found.", configPath);

    var config = {};

    config.server = {
      "name": "BioXHOME",
      "username": "99:89:83:63:64:12",
      "pin": "836-36-412"
    };

    return config;
  }

  var config;
  try {
    config = JSON.parse(fs.readFileSync(configPath));
  }
  catch (err) {
    log.error("There was a problem reading your server.json file.");
    log.error("Please try pasting your server.json file here to validate it: http://jsonlint.com");
    log.error("");
    throw err;
  }

  var accessoryCount = (config.accessories && config.accessories.length) || 0;

  var username = config.server.username;
  var validMac = /^([0-9A-F]{2}:){5}([0-9A-F]{2})$/;
  if (!validMac.test(username)){
      throw new Error('Not a valid username: ' + username + '. Must be 6 pairs of colon-' +
                      'separated hexadecimal chars (A-F 0-9), like a MAC address.');
  }

  var accessoryCount = (config.accessories && config.accessories.length) || 0;
  var platformCount = (config.platforms && config.platforms.length) || 0;
  log.info("Loaded server.json with %s accessories and %s platforms.", accessoryCount, platformCount);

  log.info("---");

  return config;
}

Server.prototype._loadCachedPlatformAccessories = function() {
  var cachedAccessories = accessoryStorage.getItem("cachedAccessories");
  var platformAccessories = [];

  if (cachedAccessories) {
    for (var index in cachedAccessories) {
      var serializedAccessory = cachedAccessories[index];
      var platformAccessory = new PlatformAccessory(serializedAccessory.displayName, serializedAccessory.UUID, serializedAccessory.category);
      platformAccessory._configFromData(serializedAccessory);

      platformAccessories.push(platformAccessory);
    }
  }

  return platformAccessories;
}

Server.prototype._createBridge = function() {
  var bridgeConfig = this._config.server || {};
  return new Bridge(bridgeConfig.name || 'BioXHOME', uuid.generate("BioXHOME"));
}

Server.prototype._loadAccessories = function() {

  log.info("Loading " + this._config.accessories.length + " accessories...");

  for (var i=0; i<this._config.accessories.length; i++) {

    var accessoryConfig = this._config.accessories[i];
    var accessoryType = accessoryConfig["accessory"];
    var accessoryConstructor = this._api.accessory(accessoryType);

    if (!accessoryConstructor)
      throw new Error("Your server.json is requesting the accessory '" + accessoryType + "' which has not been published by any installed plugins.");

    var accessoryName = accessoryConfig["name"];
    var accessoryLogger = Logger.withPrefix(accessoryName);
    accessoryLogger("Initializing %s accessory...", accessoryType);
    var accessoryInstance = new accessoryConstructor(accessoryLogger, accessoryConfig);
    var accessory = this._createAccessory(accessoryInstance, accessoryName, accessoryType, accessoryConfig.uuid_base);
    this._bridge.addBridgedAccessory(accessory);
  }
}

Server.prototype._loadPlatforms = function() {

    log.info("Loading " + this._config.platforms.length + " platforms...");

    for (var i=0; i<this._config.platforms.length; i++) {

        var platformConfig = this._config.platforms[i];
        var platformType = platformConfig["platform"];
        var platformName = platformConfig["name"] || platformType;
        var platformConstructor = this._api.platform(platformType);

        if (!platformConstructor)
          throw new Error("Your server.json is requesting the platform '" + platformType + "' which has not been published by any installed plugins.");

        var platformLogger = Logger.withPrefix(platformName);
        platformLogger("Initializing %s platform...", platformType);
        var platformInstance = new platformConstructor(platformLogger, platformConfig, this._api);

        if (platformInstance.configureAccessory == undefined) {
          this._loadPlatformAccessories(platformInstance, platformLogger, platformType);
        } else {
          this._activeDynamicPlugins[platformType] = platformInstance;
        }

        if (platformInstance.configurationRequestHandler != undefined) {
          this._configurablePlatformPlugins[platformType] = platformInstance;
        }
    }
}

Server.prototype._loadDynamicPlatforms = function() {
  for (var dynamicPluginName in this._api._dynamicPlatforms) {
    if (!this._activeDynamicPlugins[dynamicPluginName] && !this._activeDynamicPlugins[dynamicPluginName.split(".")[1]]) {
      console.log("Load " + dynamicPluginName);
      var platformConstructor = this._api._dynamicPlatforms[dynamicPluginName];
      var platformLogger = Logger.withPrefix(dynamicPluginName);
      var platformInstance = new platformConstructor(platformLogger, null, this._api);
      this._activeDynamicPlugins[dynamicPluginName] = platformInstance;

      if (platformInstance.configurationRequestHandler != undefined) {
          this._configurablePlatformPlugins[dynamicPluginName] = platformInstance;
      }
    }
  }
}

Server.prototype._configCachedPlatformAccessories = function() {
  for (var index in this._cachedPlatformAccessories) {
    var accessory = this._cachedPlatformAccessories[index];

    if (!(accessory instanceof PlatformAccessory)) {
      console.log("Unexpected Accessory!");
      continue;
    }

    var fullName = accessory._associatedPlugin + "." + accessory._associatedPlatform;
    var platformInstance = this._activeDynamicPlugins[fullName];

    if (!platformInstance) {
      platformInstance = this._activeDynamicPlugins[accessory._associatedPlatform];
    }

    if (platformInstance) {
      platformInstance.configureAccessory(accessory);
    } else {
      console.log("Failed to find plugin to handle accessory " + accessory.displayName);
    }

    accessory._prepareAssociatedHAPAccessory();
    this._bridge.addBridgedAccessory(accessory._associatedHAPAccessory);
  }
}

Server.prototype._loadPlatformAccessories = function(platformInstance, log, platformType) {
  this._asyncCalls++;
  platformInstance.accessories(once(function(foundAccessories){
      this._asyncCalls--;

      for (var i = 0; i < foundAccessories.length; i++) {
          var accessoryInstance = foundAccessories[i];
          var accessoryName = accessoryInstance.name;
          log("Initializing platform accessory '%s'...", accessoryName);
          var accessory = this._createAccessory(accessoryInstance, accessoryName, platformType, accessoryInstance.uuid_base);
          this._bridge.addBridgedAccessory(accessory);
      }

      if (this._asyncCalls === 0 && !this._asyncWait)
        this._publish();
  }.bind(this)));
}

Server.prototype._createAccessory = function(accessoryInstance, displayName, accessoryType, uuid_base) {

  var services = accessoryInstance.getServices();

  if (!(services[0] instanceof Service)) {
    return AccessoryLoader.parseAccessoryJSON({
      displayName: displayName,
      services: services
    });
  }
  else {
    var accessoryUUID = uuid.generate(accessoryType + ":" + (uuid_base || displayName));
    var accessory = new Accessory(displayName, accessoryUUID);
    if (accessoryInstance.identify)
      accessory.on('identify', function(paired, callback) { accessoryInstance.identify(callback); });

    services.forEach(function(service) {
      if (service instanceof Service.AccessoryInformation) {
        var existingService = accessory.getService(Service.AccessoryInformation);
        var manufacturer = service.getCharacteristic(Characteristic.Manufacturer).value;
        var model = service.getCharacteristic(Characteristic.Model).value;
        var serialNumber = service.getCharacteristic(Characteristic.SerialNumber).value;
        var firmwareRevision = service.getCharacteristic(Characteristic.FirmwareRevision).value;
        var hardwareRevision = service.getCharacteristic(Characteristic.HardwareRevision).value;

        if (manufacturer) existingService.setCharacteristic(Characteristic.Manufacturer, manufacturer);
        if (model) existingService.setCharacteristic(Characteristic.Model, model);
        if (serialNumber) existingService.setCharacteristic(Characteristic.SerialNumber, serialNumber);
        if (firmwareRevision) existingService.setCharacteristic(Characteristic.FirmwareRevision, firmwareRevision);
        if (hardwareRevision) existingService.setCharacteristic(Characteristic.HardwareRevision, hardwareRevision);
      }
      else {
        accessory.addService(service);
      }
    });

    return accessory;
  }
}

Server.prototype._handleRegisterPlatformAccessories = function(accessories) {
  var hapAccessories = [];
  for (var index in accessories) {
    var accessory = accessories[index];
    accessory._prepareAssociatedHAPAccessory();
    hapAccessories.push(accessory._associatedHAPAccessory);
    this._cachedPlatformAccessories.push(accessory);
  }

  this._bridge.addBridgedAccessories(hapAccessories);
  this._updateCachedAccessories();
}

Server.prototype._handleUpdatePlatformAccessories = function(accessories) {
  this._updateCachedAccessories();
}

Server.prototype._handleUnregisterPlatformAccessories = function(accessories) {
  var hapAccessories = [];
  for (var index in accessories) {
    var accessory = accessories[index];

    if (accessory._associatedHAPAccessory) {
      hapAccessories.push(accessory._associatedHAPAccessory);
    }

    for (var targetIndex in this._cachedPlatformAccessories) {
      var existing = this._cachedPlatformAccessories[targetIndex];
      if (existing.UUID === accessory.UUID) {
        this._cachedPlatformAccessories.splice(targetIndex, 1);
        break;
      }
    }
  }

  this._bridge.removeBridgedAccessories(hapAccessories);
  this._updateCachedAccessories();
}

Server.prototype._handlePublishCameraAccessories = function(accessories) {
  var accessoryPin = (this._config.server || {}).pin || "836-36-412";

  for (var index in accessories) {
    var accessory = accessories[index];

    accessory._prepareAssociatedHAPAccessory();
    var hapAccessory = accessory._associatedHAPAccessory;
    var advertiseAddress = mac.generate(accessory.UUID);

    if (this._publishedCameras[advertiseAddress]) {
      throw new Error("Camera accessory " + accessory.displayName + " experienced an address collision.");
    } else {
      this._publishedCameras[advertiseAddress] = accessory;
    }

    (function(name){
      hapAccessory.on('listening', function(port) {
  
          log.info("%s is running in the port %s.", name, port);
      })
    })(accessory.displayName);

    hapAccessory.publish({
      username: advertiseAddress,
      pincode: accessoryPin,
      category: accessory.category
    }, this._allowInsecureAccess);
  }
}

Server.prototype._updateCachedAccessories = function() {
  var serializedAccessories = [];

  for (var index in this._cachedPlatformAccessories) {
    var accessory = this._cachedPlatformAccessories[index];
    serializedAccessories.push(accessory._dictionaryPresentation());
  }

  accessoryStorage.setItemSync("cachedAccessories", serializedAccessories);
}

Server.prototype._handleNewConfig = function(type, name, replace, config) {
  if (type === "accessory") {
    if (!this._config.accessories) {
      this._config.accessories = [];
    }

    if (!replace) {
      this._config.accessories.push(config);
    } else {
      var targetName;
      if (name.indexOf('.') !== -1) {
        targetName = name.split(".")[1];
      }
      var found = false;
      for (var index in this._config.accessories) {
        var accessoryConfig = this._config.accessories[index];
        if (accessoryConfig.accessory === name) {
          this._config.accessories[index] = config;
          found = true;
          break;
        }

        if (targetName && (accessoryConfig.accessory === targetName)) {
          this._config.accessories[index] = config;
          found = true;
          break;
        }
      }

      if (!found) {
        this._config.accessories.push(config);
      }
    }
  } else if (type === "platform") {
    if (!this._config.platforms) {
      this._config.platforms = [];
    }

    if (!replace) {
      this._config.platforms.push(config);
    } else {
      var targetName;
      if (name.indexOf('.') !== -1) {
        targetName = name.split(".")[1];
      }

      var found = false;
      for (var index in this._config.platforms) {
        var platformConfig = this._config.platforms[index];
        if (platformConfig.platform === name) {
          this._config.platforms[index] = config;
          found = true;
          break;
        }

        if (targetName && (platformConfig.platform === targetName)) {
          this._config.platforms[index] = config;
          found = true;
          break;
        }
      }

      if (!found) {
        this._config.platforms.push(config);
      }
    }
  }

  var serializedConfig = JSON.stringify(this._config, null, '  ');
  var configPath = User.configPath();
  fs.writeFileSync(configPath, serializedConfig, 'utf8');
}

Server.prototype._printPin = function(pin) {
  console.log("Pin code for iOS device to pair with BioXHOME:");
  console.log(chalk.black.bgWhite("                       "));
  console.log(chalk.black.bgWhite("    ┌────────────┐     "));
  console.log(chalk.black.bgWhite("    │ " + pin + " │     "));
  console.log(chalk.black.bgWhite("    └────────────┘     "));
  console.log(chalk.black.bgWhite("                       "));
}

Server.prototype._printSetupInfo = function() {
  console.log("Setup URI Payload:");
  console.log(this._bridge.setupURI());

  console.log("QR Code for iOS device to pair with BioXHOME:");
  qrcode.generate(this._bridge.setupURI());
}

module.exports = {
  Server: Server
}
